pallets/corporate-actions/src/ballot.rs

```diff
@@ -0,0 +1,544 @@
+
+        /// Stores each DID's votes in a given ballot.
+        ///
+        /// (CAId) => (DID) => [vote weight]
+        ///
+        /// User must enter 0 vote weight if they don't want to vote for a choice.
+        pub Votes get(fn votes): double_map hasher(blake2_128_concat) CAId, hasher(blake2_128_concat) IdentityId => Vec<(T::Balance, u16)>;
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        type Error = Error<T>;
+
+        fn deposit_event() = default;
+
+        /// Attach a corporate ballot to the CA identified by `ca_id`.
+        ///
+        /// The ballot will admit votes within `range`.
+        /// The ballot's metadata is provided by `meta`,
+        /// which includes the ballot title, the motions, their choices, etc.
+        /// See the `BallotMeta` for more.
+        ///
+        /// ## Arguments
+        /// - `origin` which must be a signer for the CAA of `ca_id`.
+        /// - `ca_id` identifies the CA to attach the ballot to.
+        /// - `range` specifies when voting starts and ends.
+        /// - `meta` specifies the ballot's metadata as aforementioned.
+        /// - `rcv` specifies whether RCV is enabled for this ballot.
+        ///
+        /// # Errors
+        /// - `UnauthorizedAsAgent` if `origin` is not `ticker`'s sole CAA (owner is not necessarily the CAA).
+        /// - `NoSuchCA` if `ca_id` does not identify an existing CA.
+        /// - `CANotNotice` if the CA is not of the `IssuerNotice` kind.
+        /// - `StartAfterEnd` if `range.start > range.end`.
+        /// - `NowAfterEnd` if `now > range.end` where `now` is the current timestamp.
+        /// - `NoRecordDate` if CA has no record date.
+        /// - `RecordDateAfterStart` if `date > range.start` where `date` is the CA's record date.
+        /// - `AlreadyExists` if there's a ballot already.
+        /// - `NumberOfChoicesOverflow` if the total choice in `meta` overflows `usize`.
+        /// - `InsufficientBalance` if the protocol fee couldn't be charged.
+        #[weight = 900_000_000]
+        pub fn attach_ballot(origin, ca_id: CAId, range: BallotRange, meta: BallotMeta, rcv: bool) {
+            // Ensure origin is CAA, that `ca_id` exists, that its a notice, and the date invariant.
+            let caa = <CA<T>>::ensure_ca_agent(origin, ca_id.ticker)?;
+            let ca = <CA<T>>::ensure_ca_exists(ca_id)?;
+            ensure!(matches!(ca.kind, CAKind::IssuerNotice), Error::<T>::CANotNotice);
+            Self::ensure_range_invariant(&ca, range)?;
+
+            // Ensure CA doesn't have a ballot yet.
+            ensure!(!Ranges::contains_key(ca_id), Error::<T>::AlreadyExists);
+
+            // Compute number-of-choices-in-motion cache.
+            let choices = Self::derive_motion_num_choices(&meta.motions)?;
+
+            // Charge protocol fee.
+            T::ProtocolFee::charge_fee(ProtocolOp::VotingAddBallot)?;
```

Comment:
> We can also rename the dispatchable something like `add_ballot()` instead of `attach_ballot()`

I've intentionally avoided `add_ballot` as it suggests that you can have more than one ballot, but the relation between CA => Ballot is zero-to-one.

Will fix the rest by adding another ProtocolOp.